### PR TITLE
Tickets/DM-8973

### PR DESCRIPTION
**Pull request by natelust**

### CODE REVIEW

File: python/lsst/afw/detection/SConscript [hunk 1]
```diff
@@ -1,6 +1,7 @@
 # -*- python -*-
 from lsst.sconsUtils import scripts
 scripts.BasicSConscript.pybind11(['psf',
+				  'bootprint',
```
Comment (pschella) [inline python/lsst/afw/detection/SConscript]: space

File: python/lsst/afw/detection/bootprint.cc [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <https://www.lsstcorp.org/LegalNotices/>.
+ */
+
+
+#include <cstdint>
```
Comment (pschella) [inline python/lsst/afw/detection/bootprint.cc]: newline

File: python/lsst/afw/detection/bootprint.cc [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
```
Comment (pschella) [inline python/lsst/afw/detection/bootprint.cc]: newline

File: python/lsst/afw/detection/bootprint.cc [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+#include "numpy/arrayobject.h"
+#include "ndarray/pybind11.h"
+#include "ndarray/converter.h"
```
Comment (pschella) [inline python/lsst/afw/detection/bootprint.cc]: newline

File: python/lsst/afw/detection/bootprint.cc [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    clsBootprint.def(py::init<std::shared_ptr<geom::SpanSet>,
+                              afw::table::Schema const &,
+                              geom::Box2I const &>(),
+                     "inputSpans"_a, "peakSchema"_a, "region"_a = geom::Box2I());
+
+    /* Bootprint Methods */
+    clsBootprint.def("getSpans", &Bootprint::getSpans);
+    clsBootprint.def("setSpans", &Bootprint::setSpans);
+    clsBootprint.def("getPeaks", (PeakCatalog & (Bootprint::*)()) &Bootprint::getPeaks,
+                     py::return_value_policy::reference);
```
Comment (pschella) [inline python/lsst/afw/detection/bootprint.cc]: Are you sure? What happens if the `Bootprint` object is destroyed before the return value is? Perhaps you want `reference_internal`?

File: python/lsst/afw/detection/bootprint.cc [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+                                       (Bootprint const & self)->geom::ellipses::Quadrupole {
+                                           return self.getShape();
+                                       });
+    clsBootprint.def_property_readonly("bbox",
+                                       []
+                                       (Bootprint const & self)->geom::Box2I {
+                                           return self.getBBox();
+                                       });
+
+    /* Python magic functions */
```
Comment (pschella) [inline python/lsst/afw/detection/bootprint.cc]: These are usually just in the `Operators` section.

File: src/detection/Bootprint.cc [hunk 7]
```diff
@@ -129,6 +129,32 @@ void Bootprint::removeOrphanPeaks() {
     }
 }
 
+bool Bootprint::operator==(Bootprint const & other) const {
+    /* If the peakCatalogs are not the same length the Bootprints can't be equal */
+    if (getPeaks().size() != other.getPeaks().size()) {
```
Comment (pschella) [inline src/detection/Bootprint.cc]: I assume `::size` is cheap?

File: src/detection/Bootprint.cc [hunk 8]
```diff
@@ -129,6 +129,32 @@ void Bootprint::removeOrphanPeaks() {
...
+        return false;
+    }
+    /* Check that for every peak in the PeakCatalog there is a corresponding peak
+     * in the other, and if not return false
+     */
+    for (auto const & selfPeak : getPeaks()) {
+        int counter = 0;
```
Comment (pschella) [inline src/detection/Bootprint.cc]: `bool match = false`

File: src/detection/Bootprint.cc [hunk 9]
```diff
@@ -129,6 +129,32 @@ void Bootprint::removeOrphanPeaks() {
...
+        for (auto const & otherPeak : other.getPeaks()) {
+            if (selfPeak.getI() == otherPeak.getI() &&
+                selfPeak.getF() == otherPeak.getF() &&
+                selfPeak.getPeakValue() == otherPeak.getPeakValue()) {
```
Comment (pschella) [inline src/detection/Bootprint.cc]: `match = true;`
`break;`

File: src/detection/Bootprint.cc [hunk 10]
```diff
@@ -129,6 +129,32 @@ void Bootprint::removeOrphanPeaks() {
...
+                counter += 1;
+            }
+        }
+        if (counter == 0) {
```
Comment (pschella) [inline src/detection/Bootprint.cc]: `if (!match) {`

File: src/detection/Bootprint.cc (hunk 8 above)
Comment (pschella) [inline src/detection/Bootprint.cc]: delete counter and break out early, you only care about one match right?

File: src/detection/Bootprint.cc [hunk 11]
```diff
@@ -158,7 +184,7 @@ class LegacyBootprintPersistenceHelper {
         spanY(spanSchema.addField<int>("y", "The row of the span", "pixel")),
         spanX0(spanSchema.addField<int>("x0", "First column of span (inclusive)", "pixel")),
         spanX1(spanSchema.addField<int>("x1", "Second column of span (inclusive)", "pixel"))
-        {}
+        {spanSchema.getCitizen().markPersistent();}
```
Comment (pschella) [inline src/detection/Bootprint.cc]: Isn't this making it liver forever?

File: src/detection/Bootprint.cc [hunk 12]
```diff
@@ -168,6 +194,7 @@ std::pair<afw::table::Schema&, table::Key<int>&> spanSetPersistenceHelper() {
     if (initialize) {
         idKey = spanSetIdSchema.addField<int>("id", "id of the SpanSet catalog");
         initialize = false;
+        spanSetIdSchema.getCitizen().markPersistent();
```
Comment (pschella) [inline src/detection/Bootprint.cc]: same here

File: src/geom/SpanSet.cc [hunk 13]
```diff
@@ -761,7 +761,9 @@ namespace {
             spanSetSchema(),
             spanY(spanSetSchema.addField<int>("y", "The row of the span", "pixel")),
             spanX0(spanSetSchema.addField<int>("x0", "First column of span (inclusive)", "pixel")),
-            spanX1(spanSetSchema.addField<int>("x1", "Second column of span (inclusive)", "pixel")) {}
+            spanX1(spanSetSchema.addField<int>("x1", "Second column of span (inclusive)", "pixel")) {
+                spanSetSchema.getCitizen().markPersistent();
```
Comment (pschella) [inline src/geom/SpanSet.cc]: Again living forever?

File: tests/testBootprint.py [hunk 14]
```diff
@@ -0,0 +1,246 @@
...
+
+    def testIsHeavy(self):
+        self.assertFalse(self.footprint.isHeavy)
+
+    def testGetSetSpans(self):
+        '''
+        Test getting and setting the SpanSet member of the Bootprint with both
+        the getters and setters and the python property accessor
+        '''
+        self.assertTrue(self.footprint.getSpans() == self.spans)
```
Comment (pschella) [inline tests/testBootprint.py]: `assertEquals`?

File: tests/testBootprint.py [hunk 15]
```diff
@@ -0,0 +1,246 @@
...
+        self.footprint.spans = self.spans
+        self.assertTrue(self.footprint.spans == self.spans)
+
+    def testPeakFunctionality(self):
+        newSchema = afwDet.PeakTable.makeMinimalSchema()
+        newField = afwTable.Field_I("extra", "doc", "na")
+        newSchema.addField(newField)
+        self.footprint.setPeakSchema(newSchema)
+        names = self.footprint.getPeaks().getSchema().getNames()
+        self.assertTrue("extra" in names)
```
Comment (pschella) [inline tests/testBootprint.py]: `assertIn`?

File: tests/testBootprint.py [hunk 16]
```diff
@@ -0,0 +1,246 @@
...
+        self.spanRad = 4
+        self.regionRad = 10
+        self.spans = afwGeom.SpanSet.spanSetFromShape(self.spanRad,
+                                                      afwGeom.Stencil.BOX)
+        minPoint = afwGeom.Point2I(-self.regionRad, -self.regionRad)
+        maxPoint = afwGeom.Point2I(self.regionRad, self.regionRad)
+        self.region = afwGeom.Box2I(minPoint, maxPoint)
+        self.schema = afwDet.PeakTable.makeMinimalSchema()
+        # Run the the constructors test to ensure the Bootprints are setUp
+        self.testConstructors()
```
Comment (pschella) [inline tests/testBootprint.py]: Isn't this now run twice?

File: tests/testBootprint.py (hunk 14 above)
Comment (pschella) [inline tests/testBootprint.py]: here and everywhere below.

File: tests/testBootprint.py [hunk 17]
```diff
@@ -0,0 +1,246 @@
...
+        '''
+        Test getting and setting the SpanSet member of the Bootprint with both
+        the getters and setters and the python property accessor
+        '''
+        self.assertTrue(self.footprint.getSpans() == self.spans)
+        self.assertTrue(self.footprint.spans == self.spans)
+        tempSpanSet = afwGeom.SpanSet.spanSetFromShape(2, afwGeom.Stencil.BOX)
+        self.footprint.setSpans(tempSpanSet)
+        self.assertTrue(self.footprint.spans == tempSpanSet)
+        # reset back to origional with property
```
Comment (pschella) [inline tests/testBootprint.py]: original :)

Comment (pschella) [inline tests/testBootprint.py]: Can tests run in parallel? If so this will be a problem.

File: tests/testBootprint.py [hunk 18]
```diff
@@ -0,0 +1,246 @@
...
+        sortKey = self.footprint.peaks.getSchema()['peakValue'].asKey()
+        self.footprint.sortPeaks(sortKey)
+        for i, peak in enumerate(self.footprint.peaks):
+            self.assertEqual(peak['i_x'], i)
+
+        # Test that peaks outside the Bootprint are removed
+        self.footprint.removeOrphanPeaks()
+        self.assertEqual(len(self.footprint.peaks), 2)
+        for peak in self.footprint.peaks:
+            self.assertNotEqual(peak['i_x'], 1)
```
Comment (pschella) [inline tests/testBootprint.py]: reset the peaks?

Comment (pschella) [inline tests/testBootprint.py]: perhaps add a general `self.resetFootprint` function and always call that at the beginning? Or always construct a new one?

File: tests/testBootprint.py [hunk 19]
```diff
@@ -0,0 +1,246 @@
...
+
+    def testGeometry(self):
+        # Move the base footprint by 2 in x and 2 in y
+        offset = 2
+        self.footprint.shift(offset, offset)
+        # verify that this shifts the center from 0,0 as the default
+        # constructed footprint has
+        center = self.footprint.getCentroid()
+        self.assertEqual(center.getX(), offset)
+        self.assertEqual(center.getY(), offset)
```
Comment (pschella) [inline tests/testBootprint.py]: Use different offset for `x` and `y`. And perhaps also a negative one?

File: python/lsst/afw/detection/bootprint.cc [hunk 20]
```diff
@@ -0,0 +1,123 @@
...
+                                     (Bootprint const & self,
+                                      geom::Point2I const & point)->bool {
+                                          return self.contains(point);
+                                      });
+    clsBootprint.def("__eq__", []
+                               (Bootprint const & self,
+                                Bootprint const & other)->bool {
+                                    return self == other;
+                                }, py::is_operator());
+}
```
Comment (natelust) [inline python/lsst/afw/detection/bootprint.cc]: Note to self, need to return mod here

File: tests/testBootprint.py [hunk 21]
```diff
@@ -0,0 +1,246 @@
...
+        bBox = self.footprint.getBBox()
+        self.assertEqual(bBox.getMinX(), -self.spanRad)
+        self.assertEqual(bBox.getMinY(), -self.spanRad)
+        self.assertEqual(bBox.getMaxX(), self.spanRad)
+        self.assertEqual(bBox.getMaxY(), self.spanRad)
+
+        # Test the point membership in a Bootprint
+        memberPoint = afwGeom.Point2I(1, 1)
+        self.assertTrue(self.footprint.contains(memberPoint))
+        self.assertTrue(memberPoint in self.footprint)
```
Comment (pschella) [inline tests/testBootprint.py]: Also test non membership.

File: tests/testBootprint.py (hunk 16 above)
Comment (natelust) [inline tests/testBootprint.py]: It is run twice for the testConstructors test. All tests only run themselves, setup and tearDown. I could separate it out into a not test function, or have the code sucked into the setUp, but then I would not have a test that clearly tests the constructors for the benefit of future readers.

File: tests/testBootprint.py (hunk 17 above)
Comment (natelust) [inline tests/testBootprint.py]: no, Each test is in its own environment.

File: tests/testBootprint.py (hunk 15 above)
Comment (natelust) [inline tests/testBootprint.py]: I talked with Fred about this, and there seemed to be no single place to look at all the assert statements, so in lie of wasting more time, I just coded something that worked.

File: tests/testBootprint.py (hunk 18 above)
Comment (natelust) [inline tests/testBootprint.py]: a new one is constructed at the beginning of each test function by the setUp function

File: tests/testBootprint.py (hunk 19 above)
Comment (natelust) [inline tests/testBootprint.py]: negative is a good idea, but why the need for different numbers?

File: python/lsst/afw/detection/SConscript (hunk 1 above)
Comment (natelust) [inline python/lsst/afw/detection/SConscript]: fixed, was actually a mixed tab issue vim inserted :(

File: python/lsst/afw/detection/bootprint.cc (hunk 2 above)
Comment (natelust) [inline python/lsst/afw/detection/bootprint.cc]: added

File: python/lsst/afw/detection/bootprint.cc (hunk 3 above)
Comment (natelust) [inline python/lsst/afw/detection/bootprint.cc]: added

File: python/lsst/afw/detection/bootprint.cc (hunk 4 above)
Comment (natelust) [inline python/lsst/afw/detection/bootprint.cc]: added

File: python/lsst/afw/detection/bootprint.cc (hunk 5 above)
Comment (natelust) [inline python/lsst/afw/detection/bootprint.cc]: changed

File: python/lsst/afw/detection/bootprint.cc (hunk 6 above)
Comment (natelust) [inline python/lsst/afw/detection/bootprint.cc]: I didn't have any other operators so I renamed this section to operators

File: python/lsst/afw/detection/bootprint.cc (hunk 20 above)
Comment (natelust) [inline python/lsst/afw/detection/bootprint.cc]: I returned mod.ptr()

File: tests/testBootprint.py (hunk 19 above)
Comment (natelust) [inline tests/testBootprint.py]: I changed it anyway as there was no reason not to, and it introduced more variation in behavior to test

File: tests/testBootprint.py (hunk 21 above)
Comment (natelust) [inline tests/testBootprint.py]: added, and changed to assetIn / assertNotIn

File: tests/testBootprint.py (hunk 18 above)
Comment (pschella) [inline tests/testBootprint.py]: ok

File: tests/testBootprint.py (hunk 16 above)
Comment (pschella) [inline tests/testBootprint.py]: ok

File: src/detection/Bootprint.cc (hunk 7 above)
Comment (natelust) [inline src/detection/Bootprint.cc]: Yes, it is just calling size on a vector ultimately

File: src/detection/Bootprint.cc (hunk 8 above)
Comment (natelust) [inline src/detection/Bootprint.cc]: yes this is a much smarter way to do this

File: src/detection/Bootprint.cc (hunk 11 above)
Comment (natelust) [inline src/detection/Bootprint.cc]: This was discussed in person, but leaving a not here for future readers. This is a standard part of the persistence layer. This basically just keeps Citizen from mistakenly thinking memory is leaking as a variable is static and does not get cleared at the time the memory examiner expects it to.

**Pull request merged**